File: star_polygon.py
import math


class StarPolygon:

    def __init__(self, val_n: int, val_step: int):
        self.n = val_n
        self.step = val_step
        self.loops = 1
        if self.n < self.step:
            raise ValueError('n must be greater than step')

        self.points_position = self.calculate_points_position()
        self.sides = self.calc_sides_paths()
        self.sides_position = self.calc_sides_position()

        self.scaler = 1
        self.transform = (0, 0)

    def __repr__(self):
        return f'StarPolygon({self.n}, {self.step})'

    def calculate_points_position(self):
        points = []
        for i in range(0, self.n):
            x = 1 * math.cos(2 * math.pi * i / self.n)
            y = 1 * math.sin(2 * math.pi * i / self.n)
            points.append((x, y))

        return points

    def calc_sides_paths(self):
        points_number = {i: 0 for i in range(0, self.n)}
        sides = []
        n = 0
        self.loops = 1
        while len(points_number) > 0:
            new_n = (n + self.step) % self.n
            if n not in points_number.keys():
                n = list(points_number.keys())[0]
                self.loops += 1
                continue
            sides.append((n, new_n))
            points_number.pop(n)
            n = new_n

        return sides

    def calc_sides_position(self):
        points = self.points_position
        sides = self.sides
        sides_position = []
        for side in sides:
            sides_position.append((points[side[0]], points[side[1]]))

        return sides_position

    def get_sides_by_loops(self):
        number_of_sides = len(self.sides)
        number_per_loop = number_of_sides // self.loops
        formatted_output = []
        print(number_per_loop)
        for loop in range(0, self.loops):
            formatted_output.append([])
            for item in range(0, number_per_loop):
                formatted_output[loop].append(self.sides[item + loop * number_per_loop])

        return formatted_output

    def get_points_real_position(self):
        points = self.points_position

        points_by_scale = [(a*self.scaler + self.transform[0], b*self.scaler + self.transform[1]) for a, b in points]

        return points_by_scale

    def get_sides_real_position(self):
        sides = self.sides_position

        sides_by_scale = [((a[0]*self.scaler + self.transform[0], a[1]*self.scaler + self.transform[1]),
                           (b[0]*self.scaler + self.transform[0], b[1]*self.scaler + self.transform[1])) for a, b in sides]

        return sides_by_scale



if __name__ == '__main__':
    star = StarPolygon(7, 2)
    print(star.get_sides_by_loops())
    star1 = StarPolygon(7, 5)
    print(star1.get_sides_by_loops())

File: web/js/starPolygon.js
class StarPolygon {
    constructor(val_n, val_step) {
        this.n = val_n;
        this.step = val_step;
        this.loops = 1;
        if (this.n < this.step) {
            throw new Error('n must be greater than step');
        }

        this.points_position = this.calculatePointsPosition();
        this.sides = this.calcSidesPaths();
        this.sides_position = this.calcSidesPosition();

        this.scaler = 1;
        this.transform = [0, 0];
    }

    toString() {
        return `StarPolygon(${this.n}, ${this.step})`;
    }

    calculatePointsPosition() {
        let points = [];
        for (let i = 0; i < this.n; i++) {
            let x = 1 * Math.cos(2 * Math.PI * i / this.n);
            let y = 1 * Math.sin(2 * Math.PI * i / this.n);
            points.push([x, y]);
        }

        return points;
    }

    calcSidesPaths() {
        let pointsNumber = {};
        for (let i = 0; i < this.n; i++) {
            pointsNumber[i] = 0;
        }

        let sides = [];
        let n = 0;
        this.loops = 1;
        while (Object.keys(pointsNumber).length > 0) {
            let newN = (n + this.step) % this.n;
            if (!(n in pointsNumber)) {
                console.log("Loop");
                n = Number(Object.keys(pointsNumber)[0]);
                this.loops++;
                continue;
            }
            sides.push([n, newN]);
            delete pointsNumber[n];
            n = newN;
        }

        return sides;
    }

    calcSidesPosition() {
        let points = this.points_position;
        let sides = this.sides;
        let sidesPosition = [];
        for (const side of sides) {
            sidesPosition.push([points[side[0]], points[side[1]]]);
        }

        return sidesPosition;
    }

    getSidesByLoops() {
        let numberOfSides = this.sides.length;
        let numberPerLoop = Math.floor(numberOfSides / this.loops);
        let formattedOutput = [];
        for (let loop = 0; loop < this.loops; loop++) {
            formattedOutput.push([]);
            for (let item = 0; item < numberPerLoop; item++) {
                formattedOutput[loop].push(this.sides[item + loop * numberPerLoop]);
            }
        }

        return formattedOutput;
    }

    getPointsRealPosition() {
        let points = this.points_position;
        let pointsByScale = points.map(([a, b]) => [a * this.scaler + this.transform[0], b * this.scaler + this.transform[1]]);
        return pointsByScale;
    }

    getSidesRealPosition() {
        let sides = this.sides_position;
        let sidesByScale = sides.map(([a, b]) => ([
            [a[0] * this.scaler + this.transform[0], a[1] * this.scaler + this.transform[1]],
            [b[0] * this.scaler + this.transform[0], b[1] * this.scaler + this.transform[1]]
        ]));
        return sidesByScale;
    }

    getSidesByLoopsRealPosition() {
        let sides = this.getSidesRealPosition();
        let splitParts = Math.ceil(sides.length / this.loops);
        let splittedSides = [];
        for (let i = 0; i < sides.length; i += splitParts) {
            splittedSides.push(sides.slice(i, i + splitParts));
        }
        return splittedSides;
    }

    getPointsByLoopsRealPosition() {
        let points = this.getPointsRealPosition();
        let splitParts = Math.ceil(points.length / this.loops);
        let splittedPoints = [];
        for (let i = 0; i < points.length; i += splitParts) {
            splittedPoints.push(points.slice(i, i + splitParts));
        }
        return splittedPoints;
    }
}
